src/lib/logical_query_plan/abstract_lqp_node.cpp

```diff
@@ -19,14 +19,22 @@ class TableStatistics;
 
 AbstractLQPNode::AbstractLQPNode(LQPNodeType node_type) : _type(node_type) {}
 
-std::shared_ptr<AbstractLQPNode> AbstractLQPNode::deep_copy() const {
-  auto copied_left_child = left_child() ? left_child()->deep_copy() : std::shared_ptr<AbstractLQPNode>();
-  auto copied_right_child = this->right_child() ? this->right_child()->deep_copy() : std::shared_ptr<AbstractLQPNode>();
+std::shared_ptr<AbstractLQPNode> AbstractLQPNode::deep_copy(
+    std::shared_ptr<std::map<std::shared_ptr<const AbstractLQPNode>, std::shared_ptr<AbstractLQPNode>>> previous_copies)
+    const {
+  if (auto it = previous_copies->find(shared_from_this()); it != previous_copies->cend()) {
+    return it->second;
+  }
+
+  auto copied_left_child = left_child() ? left_child()->deep_copy(previous_copies) : nullptr;
+  auto copied_right_child = this->right_child() ? this->right_child()->deep_copy(previous_copies) : nullptr;
 
   // We cannot use the copy constructor here, because it does not work with shared_from_this()
   auto deep_copy = _deep_copy_impl(copied_left_child, copied_right_child);
-  deep_copy->set_left_child(copied_left_child);
-  deep_copy->set_right_child(copied_right_child);
+  if (copied_left_child) deep_copy->set_left_child(copied_left_child);
```

Comment:
`if (copied_left_child)` is just to avoid the unnecessary call, or would things break without it?

Reply:
It would not break, but I think that this makes the intent clearer.